From 79f24c8b193bfb614227cb0f682b47dd0eefdbf6 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Fri, 17 Jan 2025 12:07:46 +0100
Subject: [PATCH] usb: dwc2: gadget: enable the gadget as wakeup source

In order to better support wakeup in peripheral mode, mark the gadget as
a wakeup source, e.g. make it capable from udc_start() to udc_stop().
Enable the gadget as wakeup source, when requested by the host.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: Ied80fefb44ea4bd866da9f05b9361c19550966f8
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/430478
Reviewed-by: Pankaj DEV <pankaj.dev@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/usb/dwc2/core.h   |  2 ++
 drivers/usb/dwc2/gadget.c | 36 ++++++++++++++++++++++++++++++++++++
 2 files changed, 38 insertions(+)

diff --git a/drivers/usb/dwc2/core.h b/drivers/usb/dwc2/core.h
index 345926c22dbfa..75b3a7154c90f 100644
--- a/drivers/usb/dwc2/core.h
+++ b/drivers/usb/dwc2/core.h
@@ -1096,7 +1096,9 @@ struct dwc2_hsotg {
 	unsigned int srp_success:1;
 
 	struct workqueue_struct *wq_otg;
+	struct workqueue_struct *wq_gadget;
 	struct work_struct wf_otg;
+	struct work_struct wf_gadget;
 	struct timer_list wkp_timer;
 	enum dwc2_lx_state lx_state;
 	struct dwc2_gregs_backup gr_backup;
diff --git a/drivers/usb/dwc2/gadget.c b/drivers/usb/dwc2/gadget.c
index aaf43fd3b7f11..387cb8a6c399e 100644
--- a/drivers/usb/dwc2/gadget.c
+++ b/drivers/usb/dwc2/gadget.c
@@ -4638,6 +4638,8 @@ static int dwc2_hsotg_udc_start(struct usb_gadget *gadget,
 	gadget->sg_supported = using_desc_dma(hsotg);
 	dev_info(hsotg->dev, "bound driver %s\n", driver->driver.name);
 
+	device_set_wakeup_capable(&gadget->dev, true);
+
 	return 0;
 
 err:
@@ -4660,6 +4662,8 @@ static int dwc2_hsotg_udc_stop(struct usb_gadget *gadget)
 	if (!hsotg)
 		return -ENODEV;
 
+	device_set_wakeup_capable(&gadget->dev, false);
+
 	/* Exit clock gating when driver is stopped. */
 	if (hsotg->params.power_down == DWC2_POWER_DOWN_PARAM_NONE &&
 	    hsotg->bus_suspended && !hsotg->params.no_clock_gating) {
@@ -5040,6 +5044,17 @@ static void dwc2_hsotg_dump(struct dwc2_hsotg *hsotg)
 #endif
 }
 
+static void dwc2_gadget_remote_wkup_change(struct work_struct *work)
+{
+	struct dwc2_hsotg *hsotg = container_of(work, struct dwc2_hsotg, wf_gadget);
+	int ret;
+
+	dev_dbg(hsotg->dev, "lx_state: %d\n", hsotg->lx_state);
+	ret = device_set_wakeup_enable(&hsotg->gadget.dev, hsotg->lx_state == DWC2_L2);
+	if (ret)
+		dev_dbg(hsotg->dev, "device_set_wakeup_enable failed: %d\n", ret);
+}
+
 /**
  * dwc2_gadget_init - init function for gadget
  * @hsotg: Programming view of the DWC_otg controller
@@ -5117,6 +5132,13 @@ int dwc2_gadget_init(struct dwc2_hsotg *hsotg)
 		return -EINVAL;
 	}
 
+	hsotg->wq_gadget = alloc_ordered_workqueue("dwc2_gadget", 0);
+	if (!hsotg->wq_gadget) {
+		dev_err(hsotg->dev, "Failed to create workqueue\n");
+		return -ENOMEM;
+	}
+	INIT_WORK(&hsotg->wf_gadget, dwc2_gadget_remote_wkup_change);
+
 	/* setup endpoint information */
 
 	INIT_LIST_HEAD(&hsotg->gadget.ep_list);
@@ -5128,6 +5150,7 @@ int dwc2_gadget_init(struct dwc2_hsotg *hsotg)
 						     GFP_KERNEL);
 	if (!hsotg->ctrl_req) {
 		dev_err(dev, "failed to allocate ctrl req\n");
+		destroy_workqueue(hsotg->wq_gadget);
 		return -ENOMEM;
 	}
 
@@ -5156,6 +5179,10 @@ int dwc2_hsotg_remove(struct dwc2_hsotg *hsotg)
 	usb_del_gadget_udc(&hsotg->gadget);
 	dwc2_hsotg_ep_free_request(&hsotg->eps_out[0]->ep, hsotg->ctrl_req);
 
+	if (!cancel_work_sync(&hsotg->wf_gadget))
+		flush_workqueue(hsotg->wq_gadget);
+	destroy_workqueue(hsotg->wq_gadget);
+
 	return 0;
 }
 
@@ -5163,6 +5190,8 @@ int dwc2_hsotg_suspend(struct dwc2_hsotg *hsotg)
 {
 	unsigned long flags;
 
+	flush_workqueue(hsotg->wq_gadget);
+
 	if (hsotg->lx_state != DWC2_L0) {
 		hsotg->gadget_off_for_suspend = false;
 		return 0;
@@ -5449,6 +5478,8 @@ int dwc2_gadget_enter_hibernation(struct dwc2_hsotg *hsotg)
 	hsotg->gr_backup.gpwrdn = dwc2_readl(hsotg, GPWRDN);
 	dev_dbg(hsotg->dev, "Hibernation completed\n");
 
+	queue_work(hsotg->wq_gadget, &hsotg->wf_gadget);
+
 	return ret;
 }
 
@@ -5558,6 +5589,7 @@ int dwc2_gadget_exit_hibernation(struct dwc2_hsotg *hsotg,
 
 	hsotg->hibernated = 0;
 	hsotg->lx_state = DWC2_L0;
+	queue_work(hsotg->wq_gadget, &hsotg->wf_gadget);
 	dev_dbg(hsotg->dev, "Hibernation recovery completes here\n");
 
 	return ret;
@@ -5618,6 +5650,7 @@ int dwc2_gadget_enter_partial_power_down(struct dwc2_hsotg *hsotg)
 	/* Set in_ppd flag to 1 as here core enters suspend. */
 	hsotg->in_ppd = 1;
 	hsotg->lx_state = DWC2_L2;
+	queue_work(hsotg->wq_gadget, &hsotg->wf_gadget);
 
 	dev_dbg(hsotg->dev, "Entering device partial power down completed.\n");
 
@@ -5686,6 +5719,7 @@ int dwc2_gadget_exit_partial_power_down(struct dwc2_hsotg *hsotg,
 	/* Set in_ppd flag to 0 as here core exits from suspend. */
 	hsotg->in_ppd = 0;
 	hsotg->lx_state = DWC2_L0;
+	queue_work(hsotg->wq_gadget, &hsotg->wf_gadget);
 
 	dev_dbg(hsotg->dev, "Exiting device partial Power Down completed.\n");
 	return ret;
@@ -5720,6 +5754,7 @@ void dwc2_gadget_enter_clock_gating(struct dwc2_hsotg *hsotg)
 
 	hsotg->lx_state = DWC2_L2;
 	hsotg->bus_suspended = true;
+	queue_work(hsotg->wq_gadget, &hsotg->wf_gadget);
 }
 
 /*
@@ -5759,4 +5794,5 @@ void dwc2_gadget_exit_clock_gating(struct dwc2_hsotg *hsotg, int rem_wakeup)
 	/* Change to L0 state */
 	hsotg->lx_state = DWC2_L0;
 	hsotg->bus_suspended = false;
+	queue_work(hsotg->wq_gadget, &hsotg->wf_gadget);
 }
-- 
2.51.0

